source/server.go

```diff
@@ -105,14 +113,20 @@ func (s *Server) recvRecords(stream pb.SourceService_StreamServer) error {
 	}
 }
 
-func (s *Server) SendAck(position sdk.Position) error {
+func (s *Server) SendAck(ctx context.Context, position sdk.Position) error {
 	stream := s.stream.Load()
-	if stream == nil {
-		return fmt.Errorf("no stream is open")
+	for stream == nil {
+		select {
+		case <-s.streamIsSetCh:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		stream = s.stream.Load()
 	}
 	err := (*stream).Send(&pb.Ack{AckPosition: position})
 	if err != nil {
-		return fmt.Errorf("error while sending ack into stream: %w", err)
+		sdk.Logger(ctx).Err(err).Msg("error while sending ack into stream")
```

Comment:
IMO this should be a warning (we will recover, the user doesn't have to do anything), and mention in the message that we will retry.